### ISSUE TITLE
[Bug]: App hanging/crashing when browsing Episodes by date

### ISSUE THREAD

**Issue by mschumacher69** [nova-video-player/aos-AVP#605]
### Problem description

I use nova with an SMB server that auto re-scans on every app launch. This used to work flawlessly in the past, but not anymore.

Now, whenever I browse Episodes by date, I cannot play anything unless it finishes a scanning cycle without any changes to the database because it crashes if I visit Episodes by date while it's updating the library. For example if I try to play a file while the database is being updated, it circles, even the file info such as the subtitles and media info would be circling. And if I access the network section during this time, I wouldn't see my indexed folders, even though the SMB itself is there and accessible. And if I keep browsing under Episodes by date while it's refreshing the library, the app crashes to a blue screen and restarts.

Any idea what might be causing this?

### Steps to reproduce the issue

1. Start Nova.
2. Go to Episodes by date while it's scanning and try to play a file or just browse in there or sit on that view without doing anything.
3. If you manage to play a file before it crashes, it would circle forever or crash into a blue screen and restart and if you just browse in there or just stay in this view and don't do anything, it would crash into a blue screen and restart.
4. You wouldn't be able to go to Episodes by date without a crash unless you wait for it to finish scanning and then go to Episodes by date. The successful visit would always have to be after a blank scan that doesn't add/remove anything from the library. So for example, if it scans and adds/removes something from the library, browsing Episodes by date after the scan would crash Nova. Only the next visit after a scan would be successful as long as the last scan didn't add/remove anything from the library.

### Expected behavior

You should be able to browse Episodes by date and play files without the app crashing.

### Your phone/tablet/androidTV model

Sony Bravia Android TV and Chromecast with Google TV

### Operating system version

Android 9 / Android 12

### Application version and app store

Latest Version from the Play Store and latest version from GitHub

### Debug logs

Not sure how to capture logs on Android TV.

**Comment by courville** [comment]
Hi, have you been using the v6.0 of nova and then got back to google play release one?
It would help if you can get me some logs of the crash.
Since I have not touched for a long time the google play release, it could be an external factor.
Did you get an update of your TV firmware?
This is not something that has been reported before.
To capture logs please follow https://github.com/nova-video-player/aos-AVP/wiki/Debugging and do it via the network from a PC.
EDIT: anyway I would recommend to move to v6.x of nova that should be release soon and start from a fresh install. If there is a bug it could have been fixed in this release and since this is the only version targeting API31, only this version can be released on Google Play for a bug fix.

**Comment by courville** [comment]
To capture logs please follow https://github.com/nova-video-player/aos-AVP/wiki/Debugging and do it via the network from a PC.

**Comment by mschumacher69** [comment]
Hey, no I never upgraded to v6 and then downgraded back to v5, I've always been on the latest google play release.

Ok I will wait for v6 to be released on the play store and see if the issue persists after the upgrade.

**Comment by courville** [comment]
Can you please check that you have not selected always scan SMB shares at application startup (this in network share tiles rescan options).
I would really be interested to get debug logs of the crash you are encountering.
If you want a quick fix, I would also advise to install v6 before it gets release. It is supposed to be stable now with quite a large user base.

**Comment by mschumacher69** [comment]
I do have Always scan SMB shares at application startup enabled and I have enabled it on purpose since I started using Nova over 2 years ago but I only started having this issue recently.

I will try to get you debug logs when I get the chance.

What's the best way to install v6 before release? Is it on fdroid yet or should I just sideload the latest pre-release build from here?

**Comment by courville** [comment]
It is now on f-droid (fixed it recently) but I would recommend sideloading the latest github one https://github.com/nova-video-player/aos-AVP/releases/tag/v6.0.52 since it comes with another signature (the one from google play).

**Comment by mschumacher69** [comment]
Oh perfect, this means I can install both at the same time, right?

**Comment by courville** [comment]
I do not think so since it has the same package name.

**Comment by mschumacher69** [comment]
Oh ok. So if I wanna sideload the latest one from github, I have to uninstall the play store one first?

If that's the case, is there a way to backup everything and restore it after I sideload v6? Because I have a lot of scraped movies and tv shows and it might take a long time to re-scrape them all.

**Comment by courville** [comment]
It is advised to reinstall from scratch migrating to v6 because of the migration from tvdb to tmdb for series scrape.

**Comment by mschumacher69** [comment]
So here's a logcat:
[logcat.txt](https://github.com/nova-video-player/aos-AVP/files/8654693/logcat.txt)

**Comment by courville** [comment]
Thanks for the logcat, crash is the following:
```
--------- beginning of crash
04-04 18:46:53.055 23230 23268 E AndroidRuntime: FATAL EXCEPTION: pool-2-thread-1
04-04 18:46:53.055 23230 23268 E AndroidRuntime: Process: org.courville.nova, PID: 23230
04-04 18:46:53.055 23230 23268 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at androidx.loader.content.ModernAsyncTask$3.done(SourceFile:164)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at java.lang.Thread.run(Thread.java:764)
04-04 18:46:53.055 23230 23268 E AndroidRuntime: Caused by: android.database.CursorWindowAllocationException: Cursor window allocation of 2097152 bytes failed. # Open Cursors=507 (# cursors opened by this proc=507)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at android.database.CursorWindow.<init>(CursorWindow.java:136)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at android.database.CursorWindow.<init>(CursorWindow.java:114)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at android.database.AbstractWindowedCursor.clearOrCreateWindow(AbstractWindowedCursor.java:198)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:143)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:137)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at android.database.CursorWrapper.getCount(CursorWrapper.java:60)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at android.content.ContentResolver.query(ContentResolver.java:822)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at android.content.ContentResolver.query(ContentResolver.java:753)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at androidx.core.content.ContentResolverCompat.query(SourceFile:81)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at androidx.loader.content.CursorLoader.loadInBackground(SourceFile:63)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at androidx.loader.content.CursorLoader.loadInBackground(SourceFile:41)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at androidx.loader.content.AsyncTaskLoader.onLoadInBackground(SourceFile:307)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at androidx.loader.content.AsyncTaskLoader$LoadTask.doInBackground(SourceFile:60)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at androidx.loader.content.AsyncTaskLoader$LoadTask.doInBackground(SourceFile:48)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at androidx.loader.content.ModernAsyncTask$2.call(SourceFile:141)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
04-04 18:46:53.055 23230 23268 E AndroidRuntime:        ... 3 more
```
Is this with version 6 or 5?

**Comment by courville** [comment]
Ah I see this one too:
```
04-04 19:00:20.726  2979  2979 E AndroidRuntime: Process: org.courville.nova, PID: 2979
04-04 19:00:20.726  2979  2979 E AndroidRuntime: java.lang.OutOfMemoryError: pthread_create (1040KB stack) failed: Try again
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at java.lang.Thread.nativeCreate(Native Method)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at java.lang.Thread.start(Thread.java:733)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at com.archos.mediacenter.utils.trakt.TraktService.onCreate(SourceFile:1218)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at android.app.ActivityThread.handleCreateService(ActivityThread.java:3570)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at android.app.ActivityThread.access$1300(ActivityThread.java:200)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1672)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:106)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:193)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:6718)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
04-04 19:00:20.726  2979  2979 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

**Comment by mschumacher69** [comment]
Thanks for your response, this is the latest v5 from the play store.

So what's causing this issue and would updating to v6 resolve it?

**Comment by courville** [comment]
Unfortunately the two stacktraces do not provide details on where it fails in nova code.
Anyhow, there will be no other update on v5 on Google Play since it requires to target API30 now and v5 in API29.
Let's see if you can see the same issue on v6 the only possible next version on Google Play (it targets API31).

**Comment by mschumacher69** [comment]
Okay, I will wait for v6.

**Comment by mschumacher69** [comment]
Another user with a similar issue:
https://www.reddit.com/r/NovaVideoPlayer/comments/u7xn1h/nvp_and_sony_tvs/

**Comment by mschumacher69** [comment]
@courville any eta on the v6 google play release? Or shall I just sideload the github apk? Or shall I use the amazon app store or fdroid?

Which app store do you usually publish the latest version to first?

**Comment by courville** [comment]
I am solving all the instabilities seen on amazon deployment before going wider scale with Google Play. ETA 2 weeks.
I would recommend to sideload github latest v6.

**Comment by mschumacher69** [comment]
OK thanks. Is the github apk stable enough? Thing is, Nova is my main media center and I consume a lot of TV shows daily, so I cannot afford to come 1 day and find that it's not working. Like for example when you upgraded ffmpeg to 5.0.1 which broke playback and then had to revert back to 4.4.1, that would have probably cost me 1 day of playback.

So which github apks should I usually update to in order to avoid such issues? Should I avoid ones labelled as pre-release? For example now, the latest github apk is labelled "amazon release" but you just said there are instabilities on amazon deployment, so should I install this v6.0.63 amazon release or another release?

That's why I'd rather stick with an app store release (such as play store, amazon store or fdroid) that would auto update so I don't have to keep manually sideloading apks and wonder which one to sideload and which one to avoid.

**Comment by courville** [comment]
@mschumacher69 we are talking about store releases since it is published on amazon.
By instabilities, I mean existing issues that are also present on v5 Google play store since now I have more actionable information to fix these based on sentry debug backend information collected.

And I am sorry to hear about about your 1 day playback loss.

**Comment by mschumacher69** [comment]
Oh no I haven't lost any playback time yet, I'm just trying not to lose any playback time in the future.

Sorry if I might be sounding like an ass, but I don't mean any disrespect here and I appreciate all the work that you do and all the time that you put into developing Nova free of charge, I'm telling you it blows people's minds when they see it on my TV and they all swear that "this is better than Netflix"! So I wouldn't want to invite friends over to movie night to then be surprised that it's not working, which would be disappointing.

So back to sideloading the apk from github, are you saying I must always sideload the store releases and skip the pre-releases and internal/beta releases?

And let's say I sideload the "amazon release" apk today, and tomorrow you release another apk labelled "fdroid release" and then another day you release another apk labelled "google play release", should I update from one release to the other or should I always stick with the same release? In other words, if I sideload an "amazon release", should I only update to newer amazon releases or can I update to an fdroid/google play release?

I'm just confused how to proceed with updates once I sideload the github apk.

**Comment by courville** [comment]
Process: stick with github released apk for now, when google play will kick in it will update the github one (same signature).
Stability issues that I am fixing are not seen for "normal" users. I mentioned f-droid because the github release integrated changes that fixed the f-droid automatic build from source (i.e. it was not per say a specific f-droid release).

**Comment by mschumacher69** [comment]
Ok thank you.

**Comment by mjnewton** [comment]
> And I am sorry to hear about about your 1 day playback loss.

:laughing:

**Comment by mschumacher69** [comment]
@courville I'm trying to sideload the latest APK from GitHub but it's not installing. My tv is arm64-v8a and I'm sideloading the arm64-v8a APK.

I tried both with solid explorer and file commander. I also tried uninstalling the play store version.

**Comment by courville** [comment]
Are you sure that your tv is arm64? Please try arm32.

**Comment by mschumacher69** [comment]
AIDA64 says the following:
Instruction Set: 64-bit ARMv8-A (32-bit Mode)
Supported ABIs: armeabi-v7a, armeabi
Supported 32-bi ABIs: armeabi-v7a, armeabi

Does this mean I need to install Video-armeabi-v7a-API_21+-release.apk?

Anyway, I sideloaded f-droid and installed it from there. Is that ok or should I uninstall the f-droid version and sideload the github apk?

**Comment by courville** [comment]
Yes take the github armeabi-v7a one.

**Comment by mschumacher69** [comment]
Why not stay on the f-droid build? It would be easier to update through f-droid rather than having to sideload manually from github every time there's a new version.

**Comment by courville** [comment]
F-droid is fine but APK is signed by them built from source which is not compatible with my own signature and Google play one.

**Comment by mschumacher69** [comment]
Ok. I will probably not go back to google play because you said it doesn't have adult scraping.

If I install the github apk and then update to the google play version, I would lose adult scraping, right?

**Comment by courville** [comment]
Github release has adult scraping.

**Comment by mschumacher69** [comment]
Right, but you said the play store release will overwrite the github release once updated. So if I install the github release and then it updates through the play store, I would lose adult scraping, right?

**Comment by mschumacher69** [comment]
So it still crashes from time to time on v6.0.63 (latest f-droid version). It doesn't crash mid playback, it usually crashes while I'm browsing to select what I'm gonna play.

Here's a logcat:
[logcat1.txt](https://github.com/nova-video-player/aos-AVP/files/8654694/logcat1.txt)

**Comment by courville** [comment]
OK thanks for sharing the logcat, what I see is that the crash is due to OOM:
```
05-09 19:34:47.413 13257 13302 F libc    : Fatal signal 6 (SIGABRT), code -6 (SI_TKILL) in tid 13302 (RenderThread), pid 13257 (.courville.nova)
05-09 19:34:47.765 14773 14773 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
05-09 19:34:47.765 14773 14773 F DEBUG   : Build fingerprint: 'Sony/BRAVIA_ATV3_4K_PA/BRAVIA_ATV3_4K:9/PTT1.190515.001.S43/671651:user/release-keys'
05-09 19:34:47.765 14773 14773 F DEBUG   : Revision: '0'
05-09 19:34:47.765 14773 14773 F DEBUG   : ABI: 'arm'
05-09 19:34:47.765 14773 14773 F DEBUG   : pid: 13257, tid: 13302, name: RenderThread  >>> org.courville.nova <<<
05-09 19:34:47.765 14773 14773 F DEBUG   : signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
05-09 19:34:47.765 14773 14773 F DEBUG   : Abort message: 'Encountered EGL error 12291 EGL_BAD_ALLOC during rendering'
05-09 19:34:47.766 14773 14773 F DEBUG   :     r0  00000000  r1  000033f6  r2  00000006  r3  b28e4700
05-09 19:34:47.766 14773 14773 F DEBUG   :     r4  000033c9  r5  000033f6  r6  912632dc  r7  0000010c
05-09 19:34:47.766 14773 14773 F DEBUG   :     r8  91263797  r9  00000001  r10 00000001  r11 00000000
05-09 19:34:47.766 14773 14773 F DEBUG   :     ip  00000041  sp  912632c8  lr  b28700ad  pc  b2867e3a
05-09 19:34:47.882 13257 13300 E AndroidRuntime: FATAL EXCEPTION: pool-2-thread-1
05-09 19:34:47.882 13257 13300 E AndroidRuntime: Process: org.courville.nova, PID: 13257
05-09 19:34:47.882 13257 13300 E AndroidRuntime: java.lang.OutOfMemoryError: pthread_create (1040KB stack) failed: Try again
05-09 19:34:47.882 13257 13300 E AndroidRuntime:        at java.lang.Thread.nativeCreate(Native Method)
05-09 19:34:47.882 13257 13300 E AndroidRuntime:        at java.lang.Thread.start(Thread.java:733)
05-09 19:34:47.882 13257 13300 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:975)
05-09 19:34:47.882 13257 13300 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor.processWorkerExit(ThreadPoolExecutor.java:1043)
05-09 19:34:47.882 13257 13300 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1185)
05-09 19:34:47.882 13257 13300 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
05-09 19:34:47.882 13257 13300 E AndroidRuntime:        at java.lang.Thread.run(Thread.java:764)
05-09 19:34:47.966 14773 14773 F DEBUG   :
05-09 19:34:47.966 14773 14773 F DEBUG   : backtrace:
05-09 19:34:47.966 14773 14773 F DEBUG   :     #00 pc 0001ce3a  /system/lib/libc.so (abort+62)
05-09 19:34:47.966 14773 14773 F DEBUG   :     #01 pc 00006dd5  /system/lib/liblog.so (__android_log_assert+156)
05-09 19:34:47.966 14773 14773 F DEBUG   :     #02 pc 00378605  /system/lib/libhwui.so (android::uirenderer::renderthread::EglManager::swapBuffers(android::uirenderer::renderthread::Frame const&, SkRect const&)+268)
05-09 19:34:47.966 14773 14773 F DEBUG   :     #03 pc 00374231  /system/lib/libhwui.so (android::uirenderer::skiapipeline::SkiaOpenGLPipeline::swapBuffers(android::uirenderer::renderthread::Frame const&, bool, SkRect const&, android::uirenderer::FrameInfo*, bool*)+48)
05-09 19:34:47.966 14773 14773 F DEBUG   :     #04 pc 0009ae0f  /system/lib/libhwui.so (android::uirenderer::renderthread::CanvasContext::draw()+270)
05-09 19:34:47.966 14773 14773 F DEBUG   :     #05 pc 00377a1b  /system/lib/libhwui.so (_ZNSt3__110__function6__funcIZN7android10uirenderer12renderthread13DrawFrameTask11postAndWaitEvE3$_0NS_9allocatorIS6_EEFvvEEclEv$c303f2d2360db58ed70a2d0ac7ed911b+942)
05-09 19:34:47.966 14773 14773 F DEBUG   :     #06 pc 0033f00f  /system/lib/libhwui.so (android::uirenderer::WorkQueue::process()+122)
05-09 19:34:47.966 14773 14773 F DEBUG   :     #07 pc 000a3ccb  /system/lib/libhwui.so (android::uirenderer::renderthread::RenderThread::threadLoop()+178)
05-09 19:34:47.966 14773 14773 F DEBUG   :     #08 pc 0000c197  /system/lib/libutils.so (android::Thread::_threadLoop(void*)+166)
05-09 19:34:47.966 14773 14773 F DEBUG   :     #09 pc 00063bb5  /system/lib/libc.so (__pthread_start(void*)+22)
05-09 19:34:47.966 14773 14773 F DEBUG   :     #10 pc 0001e02d  /system/lib/libc.so (__start_thread+22)
```
Now the question is why do we come to this memory consumption...

**Comment by mschumacher69** [comment]
Could it be because I have a lot of shows and movies in the library? How can we troubleshoot this further?

**Comment by courville** [comment]
I do not think it is related to the number of shows or movies in the library. I will try to think of some extra logging to add in order to troubleshoot this issue.
Can you please tell me more about the exact browsing actions that lead to this behavior? Is this on a specific tile?

**Comment by mschumacher69** [comment]
Ok thanks. I usually browse Episodes by date and Movies by year and this is where it mostly happens, but I will try to monitor closely to see if this is random or only happens on that screen or when I do a certain action.

**Comment by mschumacher69** [comment]
Ok so I did some further testing today and I think it always happens while browsing under Episodes by date. Also, something strange happened today where instead of crashing, it just started showing gibberish text.

Here's another logcat:
[logcat2.txt](https://github.com/nova-video-player/aos-AVP/files/8664723/logcat2.txt)

**Comment by mschumacher69** [comment]
From my recent observations, I think this is mostly happening when I'm browsing Episodes by date.

**Comment by mschumacher69** [comment]
So yeah I think I'm only experiencing this when browsing Episodes by date.

Any idea how to troubleshoot this further?

**Comment by Pentaphon** [comment]
> So yeah I think I'm only experiencing this when browsing Episodes by date.
> 
> Any idea how to troubleshoot this further?

I'm surprised the new logger isn't capturing your crash by now. I also get a crash when doing my initial scan on v6 and just browsing around the UI while it scans but I'm not sure it's related to your crash.

**Comment by mschumacher69** [comment]
Is there a new logger? How do I use it to capture the crash?

**Comment by Pentaphon** [comment]
> Is there a new logger? How do I use it to capture the crash?

I believe they added [Sentry](https://github.com/getsentry/sentry-java) to capture data to start going after crash bugs so there's nothing you have to do but I'm surprised this issue hasn't been resolved for you.

**Comment by mschumacher69** [comment]
@courville is there anything I can do to help you troubleshoot this further?

**Comment by mschumacher69** [comment]
I'm using Nova on Google TV now and this still happens. It doesn't happen as often as on my Sony TV, but it still happens occasionally.

So the issue doesn't only affect Sony TVs.

**Comment by mschumacher69** [comment]
@courville I now know exactly when this happens, it happens when I browse episodes by date while the app is  re-scanning. I have auto-rescan when opening the application enabled, so this happens every time I open the app and also after Ambient Mode, the app seems to interpret exiting ambient mode as opening the app so it re-scans every time I exit Ambient Mode when the Ambient Mode comes on top of the app.

**Comment by mschumacher69** [comment]
@courville any chance of fixing this?

**Comment by mschumacher69** [comment]
Hey @courville, any luck replicating this?

**Comment by courville** [comment]
@mschumacher69 could you please replicate the issue while using the 6.0.80 github version? I might get crash traces on sentry crashnalytics backend and see where it happens. I get traces only with github releases.

**Comment by mschumacher69** [comment]
@courville it did happen today on 6.0.80 github version. Did you get the crash traces?

**Comment by courville** [comment]
Yes I think I have it (Android 12, Chromecast) in EpisodesByDateActivity:
```
java.util.concurrent.RejectedExecutionException: Task androidx.loader.content.ModernAsyncTask$3@c385feb rejected from java.util.concurrent.ThreadPoolExecutor@bfca148[Running, pool size = 1, active threads = 1, queued tasks = 5200, completed tasks = 108]
    at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2086)
    at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:848)
    at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1394)
    at androidx.loader.content.ModernAsyncTask.executeOnExecutor(ModernAsyncTask.java:452)
    at androidx.loader.content.AsyncTaskLoader.executePendingTask(AsyncTaskLoader.java:228)
    at androidx.loader.content.AsyncTaskLoader.dispatchOnCancelled(AsyncTaskLoader.java:241)
    at androidx.loader.content.AsyncTaskLoader$LoadTask.onCancelled(AsyncTaskLoader.java:94)
    at androidx.loader.content.ModernAsyncTask.finish(ModernAsyncTask.java:488)
    at androidx.loader.content.ModernAsyncTask$InternalHandler.handleMessage(ModernAsyncTask.java:507)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loopOnce(Looper.java:201)
    at android.os.Looper.loop(Looper.java:288)
    at android.app.ActivityThread.main(ActivityThread.java:7839)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```
For internal reference only: https://sentry.io/organizations/nova-video-player/issues/3721379768/?project=6363443&referrer=release-issue-stream#exception

**Comment by courville** [comment]
I know what it is but it does not mean that it will be simple to solve.
Related to https://github.com/nova-video-player/aos-AVP/issues/141

**Comment by courville** [comment]
OK it seems that a LinkedBlockingQueue of capacity 5200 is not enough for you...
Let's try to double it and see if it works for you.
@mschumacher69  please check this build and let me know if it helps: https://drive.google.com/file/d/1Bk5X_yNrFEAcGfeL51aL1rFBfa4bza8p/view?usp=sharing

**Comment by mschumacher69** [comment]
Great thanks, I will test it out when I get back home and let you know how it goes.

**Comment by mschumacher69** [comment]
@courville unfortunately it's still crashing when browsing episodes by date while it's still re-scanning.

**Comment by courville** [comment]
@mschumacher69, I can see the crash. Now that I have your Id it is easier to spot.
Question: does it happen only while it is still scanning?
When there is no scanning are you still able to reproduce the crash?

**Comment by mschumacher69** [comment]
@courville ok I just tested thoroughly and it only crashes if I browse episodes by date while it's scanning. If I wait for the scanning to complete and then browse episodes by date, it doesn't crash.

But while testing, I noticed that some TV series have the wrong info, so I changed info for all episodes under list episodes, but when I go from list episodes back to episodes by date, it crashes. I can replicate this every time I change episode info.

So I was able to identify 2 scenarios where episodes by date crashes:
1. When browsing episodes by date while it's scanning.
2. When browsing episodes by date after changing episode info.

So it seems that episodes by date crashes every time you browse it while it's updating (either while scanning or after changing episode info).

**Comment by courville** [comment]
OK clear thanks, I will try to reproduce on my end.

**Comment by Pentaphon** [comment]
> I just tested thoroughly and it only crashes if I browse episodes by date while it's scanning.

I've actually seen this as well on my Onn 4K with 2GB of RAM. I browsed while it scanned and it crashed so I redid the scan without touching the player and it finished just fine.

**Comment by courville** [comment]
OK I have some suggestions to test. Let's not put any limit in the LinkedBlockingQueue for now. If this works, second step will be to implement 2 VideoLoaders, one for sequential non schizophrenic episode/movie by date display and another one for the other requests (might speed things up too).
Can you please test this build and report if you still have a crash (it might be the case): https://drive.google.com/file/d/1FnlbtnR8VOBKzlxX5kctOXRoXt48sHMD/view?usp=sharing

**Comment by mschumacher69** [comment]
Ok so I tested and it didn't crash when I tried browsing episodes by date while it's scanning, but it got stuck when I backed from episodes by date and I had to press the home button on my remote to exit. This only happened the 1st time I tested. Subsequent tries went fine, it doesn't crash while browsing episodes by date while scanning and I can back out of it.

One thing I noticed though is that the rounding circle in the top right disappears after a few seconds even though it still shows as re-scanning under Network and it doesn't seem to be finishing. I tried exiting the app and re-entering several times but it still seems to be stuck on re-scanning under Network but it doesn't show the rounding circle in the top right, so I'm not sure if it's actually scanning or not.

Anyway, I will keep monitoring over the next few days and report back.

Thanks for all your hard work.

**Comment by courville** [comment]
@mschumacher69 thanks for the testing and report.
This gives a valid direction for a fix. I noticed the circle stopping and starting too (I need to see what it means: perhaps internal storage scan, then network storage scan).
I have other ideas to test as well. Perhaps this will result in another build to test.

**Comment by mschumacher69** [comment]
@courville ok so here's what's happening. I watched an episode and after I finished, it was still stuck on scanning under Network without showing the rounding circle in the top right. I tried exiting and re-entering the app several times but it was still stuck on scanning under Network but it wasn't actually scanning because it didn't add the newly downloaded episodes (this might be a one time fluke, I'll monitor if it happens again).

So I force stopped it and re-entered the app, this time it started scanning properly and showing the rounding circle in the top right. I browsed episodes by date while it's rounding and all was fine, until it detected the new episodes, this is when it froze and I had to exit it with the home button.

So now the difference is that in this test build, it doesn't crash when browsing episodes by date while scanning unless it finds new episodes. Actually it now freezes instead of crashing and restarting, but as I said, only if it finds new episodes after it finishes scanning when under episodes by date, whereas before, it used to crash and restart (not freeze) when browsing episodes by date while it's scanning even if there's no new episodes to add.

So it is an improvement, but not 100% resolved yet.

**Comment by courville** [comment]
OK trading an issue with another one... Will try to reproduce.

**Comment by courville** [comment]
Let's try this one with 5s throttling on VideoSelectionLoader and a threadpool of size 1:
https://drive.google.com/file/d/17TDo4Cbb7dFoAX4kljjhSVYSKPrgK6Bu/view?usp=sharing
Not sure it will help.
It could be interesting to wait for the app to crash following the freeze (or ANR) to get a better feel of what is stuck of you can.
P.S.: instead of providing universal builds, I could provide arm32 ones to be more inline of version numbering of your device abi compatible with google play one.

**Comment by courville** [comment]
FTR, now there is an "infinite" loader pool that are updated all at once when a new episode is detected which could be overwhelming for the device and cause the freeze. Perhaps this is the interpretation of what happens. If this is the case, I do not see any viable solution (other than blocking such heavy views while scanning).

**Comment by mschumacher69** [comment]
Ok I'll give it a try, but I have to wait until I have new episodes available. I'll also wait for the app to crash or ANR once it freezes.

Yeah you can provide arm32 builds instead of universal ones.

**Comment by courville** [comment]
BTW current fixes have been released on github as 6.0.83 pre-release with an arm32 build.

**Comment by mschumacher69** [comment]
Ok I'll test on 6.0.83 and report back, but still waiting for new episodes to be available.

I'll let you know.

**Comment by mschumacher69** [comment]
@courville ok so I got new episodes, I browsed episodes by date while it was scanning, all was fine until it detected the new episodes and froze. I waited for it as you suggested, it crashed and restarted. I tried browsing episodes by date while scanning after it restarted and it was freezing and crashing every time even though it wasn't adding new episodes. This happened around 3 times and after that it was stuck on scanning under Network without showing the rounding icon in the top right like what happened last time. So I force stopped it and tried again, it froze and crashed once while scanning in episodes by date and now it doesn't crash anymore while scanning in episodes by date.

I have a suggestion for a workaround if you don't have a fix, maybe if you set it to not refresh the episodes by date view when inside this view after scanning and only refresh it in the background when outside the episodes by date view?

**Comment by courville** [comment]
Thx for the feedback and testing.
I guess I could  `LoaderManager.getInstance(this).destroyLoader(cursorLoader.getId());` when in the `EpisodesByDate` activity. This could prevent the freeze and crash. Will experiment on it.

**Comment by courville** [comment]
Strangely enough, there is no recorded crash for Chromecast device on sentry backend these last days...

**Comment by mschumacher69** [comment]
That's strange, it crashed at least 4-5 times when I was testing yesterday.

I'm on the latest 6.0.83 from github.

**Comment by courville** [comment]
Anyway the solution is for complex views (EpisodesByDate, MoviesByGenre) is to use cursor loaders that do not update so that the overall system does not explode while scanning network shares.
Downside If there is new video file, users might have to exit and re-enter the activity for a refresh.

**Comment by mschumacher69** [comment]
Yeah I think that would be an acceptable trade off as long as we don't have to exit and re-enter the whole app for a refresh.

**Comment by courville** [comment]
OK let's experiment to `stopLoader` when result is available with `onLoadFinished` cf. https://stackoverflow.com/questions/21149917/is-it-actually-unsafe-to-call-stoploading-on-a-loader-that-is-being-managed even if this is not recommended in https://developer.android.com/reference/androidx/loader/content/Loader.html#stopLoading()
This should stop the Loader to update in case of newer videos.
@mschumacher69  could you please try this build: https://drive.google.com/file/d/1qIzVO6ihrR6Keo6hwVohB4hJe_GyFR1j/view?usp=sharing and let me know if it helps?
It will not impact the other views than the thematic ones (e.g. EpisodesByDate, MoviesByGenre etc.).

**Comment by mschumacher69** [comment]
Sure, I'll let you know once I have new episodes.

**Comment by mschumacher69** [comment]
@courville does this last change only apply to new videos or does it also apply to removing videos?

Because today I deleted episodes from the share and it crashed once in episodes by date after the deletion, do you see it in the sentry backend?

I'm still waiting for new episodes to test with new videos.

**Comment by courville** [comment]
The last change only says to make a query on the db to display the current view only once and do not refresh it if there is changes while on the current view.
Thus it should not be the cause of your crash.
Perhaps your crash is this one https://sentry.io/organizations/nova-video-player/issues/3753224570/?project=6363443&query=release%3Aorg.courville.nova%406.0.84-20221119.1915%2B2547701+error.unhandled%3Atrue&referrer=issue-stream&sort=freq
```
java.lang.ArrayIndexOutOfBoundsException: null
    at androidx.leanback.widget.CursorObjectAdapter.get(CursorObjectAdapter.java:143)
    at androidx.leanback.widget.ItemBridgeAdapter.getItemViewType(ItemBridgeAdapter.java:296)
    at androidx.recyclerview.widget.RecyclerView$Recycler.validateViewHolderForOffsetPosition(RecyclerView.java:6160)
    at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6339)
    at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6300)
    at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6296)
    at androidx.leanback.widget.GridLayoutManager.getViewForPosition(GridLayoutManager.java:1143)
    at androidx.leanback.widget.GridLayoutManager$2.createItem(GridLayoutManager.java:1672)
    at androidx.leanback.widget.SingleRow.appendVisibleItems(SingleRow.java:115)
    at androidx.leanback.widget.Grid.appendVisibleItems(Grid.java:408)
    at androidx.leanback.widget.GridLayoutManager.appendVisibleItems(GridLayoutManager.java:1992)
    at androidx.leanback.widget.GridLayoutManager.scrollDirectionPrimary(GridLayoutManager.java:2515)
    at androidx.leanback.widget.GridLayoutManager.scrollHorizontallyBy(GridLayoutManager.java:2444)
    at androidx.recyclerview.widget.RecyclerView.scrollStep(RecyclerView.java:1969)
    at androidx.recyclerview.widget.RecyclerView$ViewFlinger.run(RecyclerView.java:5476)
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1037)
    at android.view.Choreographer.doCallbacks(Choreographer.java:845)
    at android.view.Choreographer.doFrame(Choreographer.java:775)
    at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1022)
    at android.os.Handler.handleCallback(Handler.java:938)
    at android.os.Handler.dispatchMessage(Handler.java:99)
    at android.os.Looper.loopOnce(Looper.java:201)
    at android.os.Looper.loop(Looper.java:288)
    at android.app.ActivityThread.main(ActivityThread.java:7839)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```
Breadcrumbs say:
```
ui.lifecycle
{
screen: EpisodesByDateActivity, 
state: started
}
Info
17:46:54
ui.lifecycle
{
screen: VideoDetailsActivity, 
state: paused
}
Info
17:46:55
ui.lifecycle
{
screen: EpisodesByDateActivity, 
state: resumed
}
Info
17:46:55
ui.lifecycle
{
screen: VideoDetailsActivity, 
state: stopped
}
Info
17:46:55
```
Perhaps I am mixing cursors somehow transitioning from one View to another one (EpisodesByDateActivity->VideoDetailsActivity->EpisodesByDateActivity)...
Would you have a way to reproduce?

**Comment by courville** [comment]
Ok, I am able to reproduce and I will fix the issue.

**Comment by mschumacher69** [comment]
Ok great. So shall I update to 6.0.86 or do you still need me to test anything on this current test build?

**Comment by courville** [comment]
Please wait for the 6.0.87 release later tonight.

**Comment by mschumacher69** [comment]
Ok I'll wait for 6.0.87.

But something odd happened today, I deleted episodes from the share and browsed episodes by date while it's scanning to see if it's gonna freeze/crash, it didn't and the episodes didn't disappear from episodes by date (which I know is working as intended), so I exited episodes by date and re-entered to see if it's going to update and if it's going to freeze/crash, it updated and the episodes disappeared and it didn't freeze/crash.

But what happened after that is that every time I play a file, video stutters heavily and plays extremely slowly and would be out of sync with audio. I tried exiting and re-entering the app, the same issue persisted. I tried force stopping the app several times, but the app blue screened and couldn't start anymore. I had to reboot the Chromecast for the issue to clear.

This issue used to happen sometimes when the app used to crash in episodes by date.

Anyway, hopefully if the whole crashing issue is fixed in 6.0.87, I'm hoping that this other issue won't happen anymore.

**Comment by Pentaphon** [comment]
> Ok I'll wait for 6.0.87.

It's already up on github releases.

**Comment by mschumacher69** [comment]
Ok thanks, I'll give it a go and let you know.

**Comment by courville** [comment]
For your video stuttering issue, it might be that the multimedia stack of the chromecast was fubar. Will need some logs if it happens again.

**Comment by mschumacher69** [comment]
> For your video stuttering issue, it might be that the multimedia stack of the chromecast was fubar. Will need some logs if it happens again.

Ok I will let you know if the stuttering issue happens again.

But the freezing/crashing issue is still happening on 6.0.87. I browsed episodes by date while it's scanning right after I installed it (no new files or deleted files), but it froze after a while and restarted, tried episodes by date again, froze and restarted again, I then waited for it to complete the scan (waited for the rounding icon in the top right to disappear) before I go into episodes by date, it didn't crash, but I noticed that even though the rounding icon disappeared in the top right, it was still showing as scanning under Network. So I force stopped it and tried again, the same thing happened again, froze and restarted while browsing episodes by date while it's scanning, then I browsed episodes by date after the rounding icon in the top right disappeared, it didn't crash but it was also still showing as scanning under Network without showing the rounding icon in the top right and it seems to stay stuck in this state where the green rescan icon keeps rounding under Network without showing a rounding icon in the top right until I force stop the app.

The only way to get it to complete the scan is to wait for the scan to complete before browsing episodes by date.

**Comment by courville** [comment]
OK thus throttling with 1 hour delay does not help. The only way is to stop loaders according to your feedback on 6.0.85 leading to the crash when entering videoIDetailsFragment and getting back to episodesByDate fragment. I thus need to understand how to not confuse loaders/cursors. Thanks for the testing and feedback.

**Comment by mschumacher69** [comment]
@courville anything I should lookout for in 6.0.88 or did you not attempt to change anything regarding this issue in this version?

**Comment by courville** [comment]
@mschumacher69 not yet sorry (working on it still).

**Comment by mschumacher69** [comment]
Ok thanks, let me know when you need me to test a new build.

**Comment by mschumacher69** [comment]
@courville any progress on this?

**Comment by mschumacher69** [comment]
@courville any chance of fixing this? Still happening on the latest github release.

**Comment by courville** [comment]
@mschumacher69 unfortunately, not working on it since I exhausted all the ideas I had for a possible fix. Will get back on the topic but it might require migration to livedata which is another scale of magnitude of refactoring.

**Comment by mschumacher69** [comment]
@courville ok thanks, just let me know if you need me to test anything.

**Comment by mschumacher69** [comment]
@courville any progress on this?

**Comment by courville** [comment]
unfortunately not for now

**Comment by Pentaphon** [comment]
@mschumacher69 i've noticed better stability on Nova thanks to courville's past year of bugfixes resulting from the sentry logs Nova has been cataloguing. What's your experience with the most recent builds on your TV?

**Comment by mschumacher69** [comment]
It still crashes whenever I visit episodes by date while it's scanning or if there was content added/removed from the last time I visited it.

**Comment by Pentaphon** [comment]
> It still crashes whenever I visit episodes by date while it's scanning or if there was content added/removed from the last time I visited it.

Ok now I have some questions that will hopefully help courville fix this issue for you: 

1. Are you getting crashes on the latest github release as of today? (v6.2.54) 
2. What device is the SMB share on? PC with Windows? Ubuntu server? Mac? Synology NAS?
3. Have you tried using SMBJ as opposed to regular SMB on Nova? SMBJ support was detailed here: https://github.com/nova-video-player/aos-AVP/issues/835 and you can figure out how to connect using SMBJ there. Do you get crashes on BOTH SMB and SMBJ connections?

**Comment by courville** [comment]
@Pentaphon, I am able to reproduce @mschumacher69 crash. It is just not easy to fix without creating a bunch of other regressions.

**Comment by Pentaphon** [comment]
@courville understood. I simply wanted to get more info from him like the source of the SMB server to see what device he is connecting to as a possible factor and also point him towards using SMBJ as a possible workaround since his crash occurred before SMBJ was introduced. It sounds like this crash is affecting his regular usage and I wanted to see if he could work around it himself in the meantime instead of waiting for a fix.

**Comment by mschumacher69** [comment]
@Pentaphon thanks, but I don't believe my crash is related to smb because it only happens when I browse episodes by date while the database is refreshing.

**Comment by Pentaphon** [comment]
> I don't believe my crash is related to smb because it only happens when I browse episodes by date while the database is refreshing.

Are you sure it is wise to assume that? You said your setup auto-rescans so SMB is definitely being used, If I were you, I'd at least try a clean install of Nova, connect via SMBJ and see if it still happens. It's been nearly 2 years since you reported this issue, so why not just try it? Maybe you will be surprised.

**Comment by mschumacher69** [comment]
Yes I'm sure because @courville can replicate my issue and he knows what's causing it and it's related to the database refresh in the episodes by date view, not smb, but fixing it is not easy. If it were related to smb he would have told me. I have tried clean installs several times but they don't fix the issue.

Anyway, I am going to try smbj, but not to fix this issue, but to try to fix [another issue](https://github.com/nova-video-player/aos-AVP/issues/1031) I'm facing.

**Comment by mschumacher69** [comment]
To explain my issue further, you can't go to Episodes by date without a crash unless you wait for it to finish scanning and then go to Episodes by date. The successful visit would always have to be after a blank scan that doesn't add/remove anything from the library. So for example, if it scans and adds/removes something from the library, browsing Episodes by date after the scan would crash Nova. Only the next visit after a scan would be successful as long as the last scan didn't add/remove anything from the library.

**Comment by mschumacher69** [comment]
@courville I just re-read the whole thread and noticed that you did seem to fix it at one point in one of the [test builds](https://drive.google.com/file/d/1qIzVO6ihrR6Keo6hwVohB4hJe_GyFR1j/view?usp=sharing) with commit [1f3381d](https://github.com/nova-video-player/aos-Video/commit/1f3381d0af76efdb7205926998dd2d4b8d7154a3), did this commit never make it to the stable release?
